### PR TITLE
Support intermediate value plots for constrained optimization

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

This PR intends to support [PR#4851](https://github.com/optuna/optuna/pull/4851) in Optuna.
PR#4851 supported the gray coloring for infeasible solutions in the intermediate value plot.
Hence, I introduced the same feature to OptunaDashBoard.

Here is an example. 

1. Run the following code:

```python
import optuna
import matplotlib.pyplot as plt
from optuna.trial import Trial, FrozenTrial
from optuna.study import create_study
from optuna.visualization import plot_intermediate_values
from typing import Sequence

def objective_with_constraints(trial: Trial) -> float:
        trial.set_user_attr("constraint", [trial.number % 2])

        trial.report(trial.number, step=0)
        trial.report(trial.number + 1, step=1)
        return 0.0

def constraints(trial: FrozenTrial) -> Sequence[float]:
    return trial.user_attrs["constraint"]

study = create_study(sampler=optuna.samplers.NSGAIIISampler(constraints_func=constraints), storage="sqlite:///test.db")
study.optimize(objective_with_constraints, n_trials=10)
```

2. Then check the visualization:

```shell
$ npm run build:dev
$ optuna-dashboard sqlite:///$HOME/test.db
```

3. The appearance of the dashboard:

image

## What does this implement/fix? Explain your changes.

1. Color the infeasible trials with `#CCCCCC`,
2. Make the infeasible lines dashed (for color blind), and
3. Add the label saying ` (infeasible)` if the trials are infeasible and complete.

### CODE REVIEW

**Comment by not522** [comment]
@HideakiImamura Could you review this PR?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have a minor comment. PTAL.

**Comment by HideakiImamura** [comment]
I confirmed that the change of this PR produces the correct plot as follows. I used [this code](https://gist.github.com/HideakiImamura/668d6f5703c58255e54ece03cd270f2f) for verification.

Screenshot 2023-11-02 at 12 09 57

File: optuna_dashboard/ts/components/GraphIntermediateValues.tsx [hunk 1]
```diff
@@ -84,16 +84,21 @@ const plotIntermediateValue = (
     const values = trial.intermediate_values.filter(
       (iv) => iv.value !== "inf" && iv.value !== "-inf" && iv.value !== "nan"
     )
+    const isFeasible = trial.constraints.every((c) => c <= 0)
+    let name = `trial #${trial.number}`
+    if (trial.state === "Running") {
+      name += ` (running)`
+    } else {
+      name += isFeasible ? `` : ` (infeasible)`
+    }
```
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply optuna_dashboard/ts/components/GraphIntermediateValues.tsx]: Done!

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Review by keisuke-umezawa — APPROVED** [review]

**Pull request merged**